cms/src/main/java/uk/nhs/digital/apispecs/handlebars/HeadingsHyperlinksFromMarkdownHelper.java

```diff
@@ -0,0 +1,245 @@
+import java.util.stream.Stream;
+
+/**
+ * <p>
+ * 'Block' type helper which, for headings present in given Markdown, invokes
+ * the 'main' template block, providing it with a context object that contains the following properties:
+ * <ul>
+ *     <li>{@code id} - value of the {@code id} attribute of the target heading</li>
+ *     <li>{@code text} - text value of the target heading</li>
+ * </ul>
+ * <p>
+ * The original purpose of the helper is to generate hyperlinks for the side navigation panel, that would point
+ * to actual headings in the corresponding sections of the content panel.
+ * <p>
+ * Note that only top two levels of headings are taken into account and only if they are defined using the
+ * <a href="https://spec.commonmark.org/0.29/#atx-headings">ATX notation</a>. The helper dynamically
+ * determines which headings are of the 'top two levels'.
+ * <p>
+ * For example, for Markdown:
+ * <pre>
+ *     # Title
+ *     ## Overview
+ *     Some Text
+ *     ## Legal use
+ *     Some more text
+ *     ### Some level three heading
+ * </pre>
+ * <p>...invocation:
+ * <pre>
+ *     {{#headingsHyperlinksFromMarkdown markdown}}
+ *     &lt;li&gt;
+ *         &lt;a href=&quot;#{{id}}&quot;&gt;{{text}}&lt;/a&gt;
+ *     &lt;/li&gt;
+ *     {{/headingsHyperlinksFromMarkdown}}
+ * </pre>
+ * <p>...will render:</p>
+ * <pre>
+ *     &lt;li&gt;
+ *         &lt;a href=&quot;#title&quot;&gt;Title&lt;/a&gt;
+ *     &lt;/li&gt;
+ *     &lt;li&gt;
+ *         &lt;a href=&quot;#overview&quot;&gt;Overview&lt;/a&gt;
+ *     &lt;/li&gt;
+ *     &lt;li&gt;
+ *         &lt;a href=&quot;#legal-use&quot;&gt;Legal use&lt;/a&gt;
+ *     &lt;/li&gt;
+ * </pre>
+ * ...but, if heading '{@code # Title}' was to be removed from the above Markdown
+ * the following would be rendered instead:
+ * <pre>
+ *     &lt;li&gt;
+ *         &lt;a href=&quot;#overview&quot;&gt;Overview&lt;/a&gt;
+ *     &lt;/li&gt;
+ *     &lt;li&gt;
+ *         &lt;a href=&quot;#legal-use&quot;&gt;Legal use&lt;/a&gt;
+ *     &lt;/li&gt;
+ *     &lt;li&gt;
+ *         &lt;a href=&quot;#some-level-three-heading&quot;&gt;Some level three heading&lt;/a&gt;
+ *     &lt;/li&gt;
+ * </pre>
+ */
+public class HeadingsHyperlinksFromMarkdownHelper implements Helper<String> {
+
+    public static final String NAME = "headingsHyperlinksFromMarkdown";
+
+    private static final Pattern HEADING_LINE_PATTERN = Pattern.compile("^(?<hashChars>#+)\\s+(?<text>.+)$");
+    private static final Pattern TRAILING_HASHES_PATTERN = Pattern.compile("^.+(?<ending>\\s#+)$");
+
+    public static HeadingsHyperlinksFromMarkdownHelper INSTANCE = new HeadingsHyperlinksFromMarkdownHelper();
+
+
+    private HeadingsHyperlinksFromMarkdownHelper() {
+        // no-op; made private to promote the use of INSTANCE
+    }
+
+    @Override
+    public Options.Buffer apply(final String markdown, final Options options) {
+
+        try {
+            return hyperlinksRenderedFromHeadingsIn(markdown, options);
```

Comment:
maybe just personal preference on my end, but IMO single-use single-line functions are harder to follow than inline logic and dont provide valuable abstraction / maintainability / test-ability benefits. there's also a lot of nested function calls in this file. unless you feel strongly i would prefer this to be written as
```
List<HeadingModel> headingModels = headingModelsFrom(markdown);
return render(headingModels, options);
```

Reply:
Thanks! I totally get the point.

I often use this convention in 'framework' methods with generic names that I override and therefore cannot change, as a way to express what exactly is the intent of the logic invoked by said generic method.

It could easily be argued that it was an overkill here and I'm not precious about it. Fixed now in the overriding, force-pushed commit.